### DISCUSSION TITLE
CMake 是怎么工作的？ | 夜天之书

### DISCUSSION THREAD

**Discussion by giscus** [tisonkun/blog#16] [Comments]
# CMake 是怎么工作的？ | 夜天之书

关于开源软件的发布相关的内容还在构思当中，先摆烂重发部分以前讨论过的关于构建系统的文章。构建和发布密不可分，可以认为都是持续交付流水线的一环。 我会分多篇文章讨论软件发布和开源软件发布的各个方面，文章内容再加以提炼放到编撰当中的《开源指南》里。  构建系统是软件开发的重要组成部分，生产环境中的绝大多数软件都由多个组件所组成，由一系列依赖和分散的编译单元聚合而成，而自动化这些组件的集成的系统，就是

https://tisonkun.org/2022/04/15/how-cmake-works/

**Comment by PragmaTwice** [comment]
感觉现在 cmake 社区里 FetchContent 用的还蛮多的，FetchContent 是在 config 阶段下载并且把依赖 config 之后引入进来（add_subdir），不会每次 build 都拉一遍，除非依赖版本变动了才会重新拉。ExternalProject 是在 build 阶段拉，所以会有些问题，现在不如 FetchContent 流行，gtest 官方文档推荐做法已经从 ExternalProject 换成 FetchContent 了。
魔改的话感觉区别倒是不太大的样子，不管自己带 vendor 和用 FetchContent 拉自己 fork 的 repo 或者拉 upstream + patch，感觉都有一样的问题就是怎么解决和  upstream 的冲突。

**Reply by tisonkun** [reply]
有哪些典型的用 FetchContent 的例子吗？我学习一个。不瞒你说我自己写 Gradle 的时候都是对着 https://github.com/apache/iceberg 一类的软件代码仓库硬抄的。

魔改确实。我当时分析 TiFlash 和 ClickHouse 的构建设计的时候，也说了 ClickHouse 是花了大力气去精心裁剪依赖和编写 CMake 脚本的（并不，他们根本是大力出奇迹暴力魔改）。ClickHouse 基本都会 fork 一份上游从此以后就自己愣冲了，可能偶尔会 pick 一下上游对严重缺陷的修复。

**Reply by PragmaTwice** [reply]
[oneflow](https://github.com/Oneflow-Inc/oneflow/tree/master/cmake/third_party) 最近好像在把依赖从 ExternalProject 改成 FetchContent，让依赖可以在 oneflow cmake config 阶段被引入（比如 [gtest](https://github.com/Oneflow-Inc/oneflow/blob/master/cmake/third_party/googletest.cmake)， [glog](https://github.com/Oneflow-Inc/oneflow/blob/master/cmake/third_party/glog.cmake)， [json](https://github.com/Oneflow-Inc/oneflow/blob/master/cmake/third_party/json.cmake)，还有 [pybind11](https://github.com/Oneflow-Inc/oneflow/blob/master/cmake/pybind11.cmake) 等等），[我自己最近写的项目](https://github.com/PragmaTwice/proxinject/blob/master/CMakeLists.txt) 也完全用 FetchContent 管理依赖。clickhouse 好像是在用 git submodule（如果我没记错），我个人还是觉得 git submodule 有些不好用的地方，比如：不能和构建（配置如何构建某个依赖）放在一起，要维护两个地方；不能（很难）依赖 HEAD；在低版本 git 里 remove submodule 有点麻烦；不太容易配置多个源，比如你想提供一个选项让用户可以在某个 mirror 里下载依赖的源代码（特别是在中国）。当然有条件的大项目，自己管理自己的 vcpkg 依赖库（自己 fork 一份 vcpkg）或者 conan 库肯定是最好的。conan 的优点是你可以直接下载 build 好的 binary。

**Comment by ZhangYet** [comment]
勘误
> 现代 CMake 的最佳时间是牢记围绕构建目标组织声明

当作 
> 现代 CMake 的最佳实践是牢记围绕构建目标组织声明

**Reply by tisonkun** [reply]
Thanks! Fixed at ae3253e7c3b3727044c9ae284b7584d50735ba1f.

你也可以直接在右上角 GitHub Logo 跳转到原文文件进行修改。
